xbmc/pvr/channels/PVRChannelProvider.cpp

```diff
@@ -0,0 +1,544 @@
+{
+  CSingleLock lock(m_critSection);
+  if (m_strIconPath != strIconPath)
+  {
+    m_strIconPath = StringUtils::Format("%s", strIconPath.c_str());
+
+    m_bIsUserSetIcon = bIsUserSetIcon && !m_strIconPath.empty();
+
+    if (!bIsUserSetIcon)
+      m_strIconPath = StringUtils::Format("%s", GetClientIconPath().c_str());
+    else if (m_strIconPath == m_strClientIconPath)
+      bIsUserSetIcon = false;
+
+    return true;
+  }
+
+  return false;
+}
+
+std::string CPVRChannelProvider::GetClientCountry() const
+{
+  CSingleLock lock(m_critSection);
+  return m_strClientCountry;
+}
+
+bool CPVRChannelProvider::SetClientCountry(const std::string& strClientCountry)
+{
+  CSingleLock lock(m_critSection);
+  if (m_strClientCountry != strClientCountry)
+  {
+    m_strClientCountry = strClientCountry;
+
+    return true;
+  }
+
+  return false;
+}
+
+std::string CPVRChannelProvider::GetCountry() const
+{
+  CSingleLock lock(m_critSection);
+  return m_strCountry;
+}
+
+bool CPVRChannelProvider::SetCountry(const std::string& strCountry,
+                                      bool bIsUserSetCountry /* = false */)
+{
+  CSingleLock lock(m_critSection);
+  if (m_strCountry != strCountry)
+  {
+    m_strCountry = StringUtils::Format("%s", strCountry.c_str());
+
+    m_bIsUserSetCountry = bIsUserSetCountry && !m_strCountry.empty();
+
+    if (!bIsUserSetCountry)
+      m_strCountry = StringUtils::Format("%s", GetClientCountry().c_str());
+    else if (m_strCountry == m_strClientCountry)
+      bIsUserSetCountry = false;
+    return true;
+  }
+
+  return false;
+}
+
+std::string CPVRChannelProvider::GetClientLanguage() const
+{
+  CSingleLock lock(m_critSection);
+  return m_strClientLanguage;
+}
+
+bool CPVRChannelProvider::SetClientLanguage(const std::string& strClientLanguage)
+{
+  CSingleLock lock(m_critSection);
+  if (m_strClientLanguage != strClientLanguage)
+  {
+    m_strClientLanguage = strClientLanguage;
+
+    return true;
+  }
+
+  return false;
+}
+
+std::string CPVRChannelProvider::GetLanguage() const
+{
+  CSingleLock lock(m_critSection);
+  return m_strLanguage;
+}
+
+bool CPVRChannelProvider::SetLanguage(const std::string& strLanguage,
+                                      bool bIsUserSetLanguage /* = false */)
+{
+  CSingleLock lock(m_critSection);
+  if (m_strLanguage != strLanguage)
+  {
+    m_strLanguage = StringUtils::Format("%s", strLanguage.c_str());
+
+    m_bIsUserSetLanguage = bIsUserSetLanguage && !m_strLanguage.empty();
+
+    if (!bIsUserSetLanguage)
+      m_strLanguage = StringUtils::Format("%s", GetClientLanguage().c_str());
+    else if (m_strLanguage == m_strClientLanguage)
+      bIsUserSetLanguage = false;
+
+    return true;
+  }
+
+  return false;
+}
+
+void CPVRChannelProvider::UpdatePath()
+{
+  m_strPath = CPVRChannelsPath::PATH_PROVIDER_CHANNELS + m_strProviderName;
+}
+
+namespace
+{
+
+std::string GetProviderTypeLabel(PVR_CHANNEL_PROVIDER_TYPE providerType)
```

Comment:
Wouldn't this kind of things belong into UI layer, not core classs?

Reply:
I was not where this should go. You're right the UI layer would be the best place.

Reply:
Question. Where exactly in the UI layer should this go? I guess there is nowhere for it to live yet.

Reply:
CPVRGUIInfo maybe? Assuming these labels will manifest as ListItem.xyz